polymetis/polymetis/src/clients/allegro_hand_client/allegro_hand_client.cpp

```diff
@@ -0,0 +1,213 @@
+  allegro_hand_ptr_->requestStatus();
+  while (recovery_attempts < RECOVERY_MAX_TRIES) {
+    try {
+      while (true) {
+        while (!allegro_hand_ptr_->allStatesUpdated()) {
+          int finger = allegro_hand_ptr_->poll();
+          if (finger >= 0) {
+            std::copy_n(allegro_hand_ptr_->getPositions(), kNDofs,
+                        robot_state.q.begin());
+            Eigen::Map<Eigen::VectorXd> positions(robot_state.q.data(), kNDofs);
+            Eigen::Map<Eigen::VectorXd> velocities(robot_state.dq.data(),
+                                                   kNDofs);
+            velocity_filter_->filter_position(positions, finger, velocities);
+            poll_success_event.checkpoint();
+            recovery_attempts = 0; // Comms are working / restored.
+          } else if (poll_success_event.secondsSinceCheckpoint() > 3) {
+            throw std::runtime_error("No recent communication from hand.");
+          }
+        }
+        allegro_hand_ptr_->resetStateUpdateTracker();
+
+        // Compute torque components
+        updateServerCommand(robot_state, torque_commanded_);
+
+        allegro_hand_ptr_->setTorques(torque_commanded_.begin());
+      }
+    } catch (const std::exception &ex) {
+      spdlog::error("Robot is unable to be controlled: {}", ex.what());
+    }
+
+    // Attempt recovery
+    spdlog::warn("Attempting to reconnect to Allegro Hand");
+    allegro_hand_ptr_->resetCommunication();
+    allegro_hand_ptr_->setServoEnable(true);
+    allegro_hand_ptr_->requestStatus();
+    poll_success_event.checkpoint(); // reset the comm heartbeat timer
+
+    recovery_attempts++;
+  }
+  spdlog::error("Max recovery attempts exhausted.");
+}
+
+void AllegroHandTorqueControlClient::updateServerCommand(
+    /*
+     * Send robot states and receive torque command via a request to the
+     * controller server.
+     */
+    const AllegroHandState &robot_state,
+    std::array<double, NUM_DOFS> &torque_out) {
+  // Record robot states
+  for (int i = 0; i < NUM_DOFS; i++) {
+    robot_state_.set_joint_positions(i, robot_state.q[i]);
+    robot_state_.set_joint_velocities(i, robot_state.dq[i]);
+  }
+  setTimestampToNow(robot_state_.mutable_timestamp());
+
+  // Retrieve torques
+  grpc::ClientContext context;
+  long int pre_update_ns = getNanoseconds();
+  status_ = stub_->ControlUpdate(&context, robot_state_, &torque_command_);
+  long int post_update_ns = getNanoseconds();
+  if (!status_.ok()) {
+    std::string error_msg = "ControlUpdate rpc failed. ";
+    throw std::runtime_error(error_msg + status_.error_message());
+  }
+
+  robot_state_.set_prev_controller_latency_ms(
+      float(post_update_ns - pre_update_ns) / 1e6);
+
+  assert(torque_command_.joint_torques_size() == NUM_DOFS);
+  for (int i = 0; i < NUM_DOFS; i++) {
+    torque_out[i] = torque_command_.joint_torques(i);
+    robot_state_.set_prev_joint_torques_computed(
+        i, torque_command_.joint_torques(i));
+  }
+}
```

Comment:
A lot of code in here and in other places of the same file is copied from the `FrankaTorqueControlClient`. It should be ok for now to get the Allegro hand working with Polymetis asap, but we should make an issue or a task in the icebox to deduplicate by maybe having a `TorqueControlClient` base class.

Reply:
Yup!

Reply:
Created an issue: #1039